qubes-gui-client/src/buffer.rs

```diff
@@ -0,0 +1,389 @@
+            let to_write = &front[self.offset..];
+            if to_write.is_empty() {
+                self.queue.pop_front();
+                self.offset = 0;
+                continue;
+            }
+            let written_this_time = Self::write_slice(&mut self.vchan, to_write)?;
+            written += written_this_time;
+            self.offset += written_this_time;
+            if written_this_time < to_write.len() {
+                break Ok(written);
+            }
+        }
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.drain()?;
+        if !self.queue.is_empty() {
+            self.queue.push_back(buf.to_owned());
+            return Ok(());
+        }
+        assert_eq!(self.offset, 0);
+        let written = Self::write_slice(&mut self.vchan, buf)?;
+        if written != buf.len() {
+            assert!(written < buf.len());
+            self.queue.push_back(buf[written..].to_owned());
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn recv(&mut self, s: Range<usize>) -> io::Result<usize> {
+        self.vchan.recv(&mut self.buffer[s]).map_err(|e| {
+            self.state = ReadState::Error;
+            e
+        })
+    }
+
+    pub fn wait(&mut self) {
+        self.vchan.wait()
+    }
+
+    /// If there is nothing to read, return `Ok(None)` immediately; otherwise,
+    /// returns `Ok(Some(msg))` if a complete message has been read, or `Err`
+    /// if something went wrong.
+    pub fn read_header(&mut self) -> io::Result<Option<(Header, &[u8])>> {
+        self.drain()?;
+        let mut ready = self.vchan.data_ready();
+        loop {
+            if ready == 0 {
+                break Ok(None);
+            }
+            match self.state {
+                ReadState::Error => {
+                    break Err(Error::new(ErrorKind::Other, "Already in error state"))
+                }
+                ReadState::ReadingHeader if ready >= size_of::<Header>() => {
+                    let mut header = <Header as Default>::default();
+                    if self.vchan.recv(header.as_mut_bytes()).map_err(|e| {
+                        self.state = ReadState::Error;
+                        e
+                    })? != size_of::<Header>()
+                    {
+                        break Err(Error::new(
+                            ErrorKind::UnexpectedEof,
+                            "Failed to read a full message header",
+                        ));
+                    }
+                    ready -= size_of::<Header>();
+                    let untrusted_len = u32_to_usize(header.untrusted_len);
+                    match qubes_gui::msg_length_limits(header.ty) {
+                        // See below comment regarding empty messages!
+                        None if untrusted_len == 0 => continue,
+                        None => self.state = ReadState::Discard(untrusted_len),
+                        Some(max_len) if max_len.contains(&untrusted_len) => {
```

Comment:
`max_len` -> `len_range` or `allowed_range`, considering that it's a range of min and max len...